File: internal/cli/review.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/joho/godotenv"
	"github.com/spf13/cobra"

	"github.com/CrowdStrike/codestrike/internal/config"
	"github.com/CrowdStrike/codestrike/internal/review"
	"github.com/CrowdStrike/codestrike/internal/setup"
	"github.com/CrowdStrike/codestrike/internal/setup/logger"
	"github.com/CrowdStrike/codestrike/internal/tokenizer"
)

func newReviewCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "review <pr-url>",
		Short: "Run an AI review on a pull request and post the result as a comment",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			_ = godotenv.Load()

			envCfg := setup.LoadConfig()
			log := logger.New(envCfg.LogLevel)

			configFlag, err := cmd.Flags().GetString("config")
			if err != nil {
				return fmt.Errorf("reading --config flag: %w", err)
			}

			fullContext, err := cmd.Flags().GetBool("full-context")
			if err != nil {
				return fmt.Errorf("reading --full-context flag: %w", err)
			}

			persona, err := cmd.Flags().GetString("persona")
			if err != nil {
				return fmt.Errorf("reading --persona flag: %w", err)
			}

			configPath, err := config.ResolvePath(configFlag)
			if err != nil {
				return fmt.Errorf("resolving config path: %w", err)
			}

			appConfig, err := config.Load(configPath)
			if err != nil {
				return fmt.Errorf("loading config: %w", err)
			}

			if persona != "" {
				promptPath := filepath.Join(filepath.Dir(configPath), "prompts", persona+".md")
				data, err := os.ReadFile(promptPath)
				if err != nil {
					return fmt.Errorf("persona %q not found (expected at %s): %w", persona, promptPath, err)
				}
				appConfig.Review.SystemPrompt = string(data)
			}

			ref, err := review.ParsePRURL(args[0])
			if err != nil {
				return fmt.Errorf("parsing PR URL: %w", err)
			}

			deps, err := setup.Wire(cmd.Context(), envCfg, appConfig, &log, ref.Owner, ref.Repo)
			if err != nil {
				return fmt.Errorf("wiring dependencies: %w", err)
			}

			tok := tokenizer.NewForModel(appConfig.Review.Context.TokenizerModel)
			pipeline := review.NewPipeline(deps.SCMClient, deps.LLMClient, deps.AppConfig, tok, deps.Logger, review.Options{
				FullContext: fullContext,
			})

			return pipeline.Run(cmd.Context(), ref)
		},
	}

	cmd.Flags().Bool("full-context", false, "Fetch full file content for richer reviews (slower, uses more tokens)")
	cmd.Flags().String("persona", "", "Review persona — maps to a prompt file in prompts/ (e.g., security, performance)")

	return cmd
}

File: src/cli/review.ts
import * as path from 'path';
import * as fs from 'fs';
import { Command } from 'commander';
import { config as loadDotenv } from 'dotenv';
import { load as loadConfig } from '../config/config.js';
import { resolvePath } from '../config/path.js';
import { parsePrUrl } from '../review/parser.js';
import { Pipeline } from '../review/pipeline.js';
import { loadEnvConfig, wire } from '../setup.js';
import { createLogger } from '../logger.js';
import { newForModel } from '../tokenizer.js';

export function reviewCommand(): Command {
  const cmd = new Command('review')
    .description('Run an AI review on a pull request and post the result as a comment (or print it with --dry-run)')
    .argument('<pr-url>', 'Pull request URL')
    .option('--full-context', 'Fetch full file content for richer reviews (slower, uses more tokens)', false)
    .option('--persona <name>', 'Review persona — maps to a prompt file in prompts/ (e.g., security, performance)')
    .option('--dry-run', 'Print the review to stdout instead of posting it as a PR comment', false)
    .action(async (prUrl: string, opts: { fullContext: boolean; persona?: string; dryRun: boolean }) => {
      loadDotenv();

      const envCfg = loadEnvConfig();
      const logger = createLogger(envCfg.logLevel);

      const configFlag = cmd.parent?.getOptionValue('config') as string | undefined;
      const configPath = resolvePath(configFlag);
      const appConfig = loadConfig(configPath);

      if (opts.persona) {
        if (path.basename(opts.persona) !== opts.persona) {
          throw new Error(`persona "${opts.persona}" must be a plain name, not a path`);
        }
        const promptPath = path.join(path.dirname(configPath), 'prompts', opts.persona + '.md');
        const data = fs.readFileSync(promptPath, 'utf-8');
        appConfig.review.system_prompt = data;
      }

      const ref = parsePrUrl(prUrl);
      const deps = wire(envCfg, appConfig, logger, ref);

      const tok = newForModel(appConfig.review.context.tokenizer_model);
      const pipeline = new Pipeline(deps.scmClient, deps.llmClient, deps.appConfig, tok, deps.logger, {
        fullContext: opts.fullContext,
        dryRun: opts.dryRun,
      });

      const body = await pipeline.run(ref);
      if (body) console.log(body);
    });

  return cmd;
}
